RFD.md

```diff
@@ -0,0 +1,217 @@
+
+Here are the original requirements as described on the [challenge documentation page](https://github.com/gravitational/careers/blob/main/challenges/systems/challenge.md):
+
+
+### Library
+
+* Worker library with methods to start/stop/query status and get the output of a job.
+* Library should be able to stream the output of a running job.
+  * Output should be from start of process execution.
+  * Multiple concurrent clients should be supported.
+* Add resource control for CPU, Memory and Disk IO per job using cgroups.
+* Add resource isolation for using PID, mount, and networking namespaces.
+
+### API
+
+* [GRPC](https://grpc.io) API to start/stop/get status/stream output of a running process.
+* Use mTLS authentication and verify client certificate. Set up strong set of cipher suites for TLS and good crypto setup for certificates. Do not use any other authentication protocols on top of mTLS.
+* Use a simple authorization scheme.
+
+### Client
+
+* CLI should be able to connect to worker service and start, stop, get status, and stream output of a job.
+
+## Solution Details
+
+### Security
+
+A remote command execution service poses a number of challenges when it comes to security:
+
+- The system needs to have a strong authentication and authorization system in place to ensure that only the people allowed to perform remote commands would be able to do so.
+
+- The data coming through the system (commands, console output, etc) needs to be protected since it may contain sensitive information (credentials, customer data, etc).
+
+- The host system running the server needs to be protected from potentially abusive and/or malicious users.
+
+- Since the system is a multi-tenant environment, it may need some mechanisms to protect its users from noisy/malicious/abusive neighbors.
+
+#### Authentication
+
+We're going to use a dedicated certificate authority to issue server and client certificates and will use it to verify both the server and the client during the establishment of an mTLS connection to the service.
+
+##### Scope limits
+
+CA, server and client certificate generation and management is out of scope for the project. A set of example certificates and keys will be checked into the repository with the project to allow for easier development of the system.
+
+In an actual production system, certificates and keys would have to be configured using a configuration file or some other mechanism and their generation and management would depend on the underlying infrastructure managed by the operators of the service.
+
+#### Authorization
+
+Simple authorization tokens will be used by the client to gain access to the system after the initial mTLS connection is established. Each client could have its own set of access tokens and tokens could be revoked by the operators of the service if needed.
```

Comment:
Can you think of a way to do authorization without needing access tokens?

Reply:
I was initially thinking of using ssl certs themselves to provide authorization for the clients, but then looked at the requirements and interpreted them in a way that would require a separate authorization mechanism. If using ssl certs by themselves is OK (like openvpn does it, for example), I can probably do that to makes things cleaner.

Reply:
Yeah if you can use the certs you already have I think that will be cleaner. 🙂

Reply:
I'll try to do it and see how it goes.

Reply:
Any updated on this one?

Reply:
To be honest, the situation is like this: I've never done it before in Go (or in GRPC in general), but I know it is possible. (I've done it in Ruby and I've used the same approach in OpenVPN as an SRE for at least a decade). I've spent some time investigating possible implementation in GRPC and it is not immediately obvious how to do it. Given that there is no easily discoverable prior art, I'd lean towards not inventing my own approach here and just use tokens. BUT, if I have time, I could absolutely investigate it further at the end and see if I could make it work (basically need to get access to certificate metadata within a stream interceptor AFAIU).

Reply:
Ok, after some finagling with request contexts I've managed to pull the client cert out and now I have access to the client cert CN, which I can use for authorization! 🥳